### PR TITLE
feat: add DRA resourceClaims support to all DSPA components

### PR DESCRIPTION
**Pull request by hbelmiro**
## Summary

- Adds `ResourceClaims []corev1.PodResourceClaim` field to all 8 DSPA component specs (APIServer, PersistenceAgent, ScheduledWorkflow, WorkflowController, MariaDB, Minio, MLMD Envoy, MLMD GRPC)
- Wires field through deployment templates using `{{ if }}/{{ range }}` pattern consistent with existing resource fields
- Regenerated DeepCopy and CRD manifests
- 12 unit tests covering all template branches and params extraction

Resolves https://redhat.atlassian.net/browse/RHOAIENG-83734

## Summary by CodeRabbit

- **New Features**
  - Added support for configuring Kubernetes pod resource claims across API Server, persistence, workflow, database, object storage, and metadata services.
  - Claims can reference an existing resource or a resource claim template.
  - Added validation for claim names, required references, and duplicate claim entries.
  - Configured claims are now propagated to generated service deployments.

- **Bug Fixes**
  - Improved safe YAML rendering for claim names and references, including values containing quotes or newlines.

- **Tests**
  - Added coverage for claim validation, deployment generation, serialization, and multi-component configurations.

### CODE REVIEW

File: api/v1/dspipeline_types.go [hunk 1]
```diff
@@ -253,6 +261,10 @@ type ScheduledWorkflow struct {
 	CronScheduleTimezone string `json:"cronScheduleTimezone,omitempty"`
 	// Specify custom Pod resource requirements for this component.
 	Resources *ResourceRequirements `json:"resources,omitempty"`
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the Pod is allowed to start.
+	// +kubebuilder:validation:Optional
+	ResourceClaims []corev1.PodResourceClaim `json:"resourceClaims,omitempty"`
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply api/v1/dspipeline_types.go]: Done.

File: config/internal/apiserver/default/deployment.yaml.tmpl [hunk 2]
```diff
@@ -360,6 +360,18 @@ spec:
...
         {{ end }}
+      {{ if .APIServer.ResourceClaims }}
+      resourceClaims:
+      {{ range .APIServer.ResourceClaims }}
+        - name: {{ .Name }}
+          {{ if .ResourceClaimName }}
+          resourceClaimName: {{ .ResourceClaimName }}
+          {{ end }}
+          {{ if .ResourceClaimTemplateName }}
+          resourceClaimTemplateName: {{ .ResourceClaimTemplateName }}
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply config/internal/apiserver/default/deployment.yaml.tmpl]: Done.

[the remark this answers is not in the document]
Comment (hbelmiro) [reply config/internal/apiserver/default/deployment.yaml.tmpl]: K8s API server already enforces exactly-one-of at pod admission time.

**Comment by hbelmiro** [comment]
@coderabbit review

File: config/internal/ml-metadata/metadata-envoy.deployment.yaml.tmpl [hunk 3]
```diff
@@ -64,6 +64,12 @@ spec:
               port: md-envoy
             timeoutSeconds: 2
           resources:
+            {{ if .MLMD.Envoy.ResourceClaims }}
+            claims:
+              {{ range .MLMD.Envoy.ResourceClaims }}
+              - name: "{{ .Name }}"
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply config/internal/ml-metadata/metadata-envoy.deployment.yaml.tmpl]: Addressed in fa0a77dda: all three claim strings now use a locally validated PodResourceClaim type with Kubernetes DNS constraints, and every pod/container claim scalar across all eight templates is emitted through yamlString. The quote/newline rendering regression test verifies injected PodSpec fields are not created.

File: config/internal/ml-metadata/metadata-envoy.deployment.yaml.tmpl [hunk 4]
```diff
@@ -145,6 +151,18 @@ spec:
...
+      {{ range .MLMD.Envoy.ResourceClaims }}
+        - name: "{{ .Name }}"
+          {{ if .ResourceClaimName }}
+          resourceClaimName: "{{ .ResourceClaimName }}"
+          {{ end }}
+          {{ if .ResourceClaimTemplateName }}
+          resourceClaimTemplateName: "{{ .ResourceClaimTemplateName }}"
+          {{ end }}
+      {{ end }}
+      {{ end }}
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply config/internal/ml-metadata/metadata-envoy.deployment.yaml.tmpl:165]: Addressed in fa0a77dda: PodResourceClaim now has a CEL exactly-one rule, which is expanded into the CRD schema for all eight component fields. TestResourceClaimCRDValidation verifies missing and dual sources are rejected while direct and template sources are accepted; direct resourceClaimName claims do not depend on resourceClaimStatuses.

**Comment by hbelmiro** [comment]
@coderabbit review

File: api/v1/dspipeline_types.go [hunk 5]
```diff
@@ -131,6 +152,10 @@ type APIServer struct {
 	ManagedPipelines *ManagedPipelinesSpec `json:"managedPipelines,omitempty"`
 	// Specify custom Pod resource requirements for this component.
 	Resources *ResourceRequirements `json:"resources,omitempty"`
+	// ResourceClaims defines which ResourceClaims must be allocated
+	// and reserved before the Pod is allowed to start.
+	// +kubebuilder:validation:Optional
+	ResourceClaims []PodResourceClaim `json:"resourceClaims,omitempty"`
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply api/v1/dspipeline_types.go:161]: Addressed in the latest commit: all eight ResourceClaims fields now use `+listType=map` and `+listMapKey=name`, and the CRD was regenerated. Added an envtest regression covering duplicate claim names.

File: controllers/resource_claim_validation_test.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+			},
+			wantErr: "spec.apiServer.resourceClaims[0].resourceClaimTemplateName",
+		},
+	}
+
+	for idx, tt := range tests {
+		s.T().Run(tt.name, func(t *testing.T) {
+			namespace := fmt.Sprintf("resource-claim-validation-%d", idx)
+			ns := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: namespace}}
+			require.NoError(t, ctrlclient.IgnoreAlreadyExists(k8sClient.Create(ctx, ns)))
```
[the remark this answers is not in the document]
Comment (hbelmiro) [reply controllers/resource_claim_validation_test.go:97]: Addressed in the latest commit: each fixed-name validation test now registers Namespace cleanup with `t.Cleanup`, so Namespaces are deleted after the DSPA cleanup.

File: api/v1/zz_generated.deepcopy.go [hunk 7]
```diff
@@ -38,6 +38,13 @@ func (in *APIServer) DeepCopyInto(out *APIServer) {
 		*out = new(ResourceRequirements)
 		(*in).DeepCopyInto(*out)
 	}
+	if in.ResourceClaims != nil {
```
Comment (alyssacgoins) [inline api/v1/zz_generated.deepcopy.go:41]: This block is repeated a number of times - can you extract this out into a helper function?

File: controllers/config/templating.go [hunk 8]
```diff
@@ -58,7 +60,9 @@ func templateSource(r io.Reader, context interface{}) (mf.Source, error) {
 	if err != nil {
 		return mf.Slice([]unstructured.Unstructured{}), err
 	}
-	t, err := template.New("manifestTemplateDSP").Parse(string(b))
+	t, err := template.New("manifestTemplateDSP").Funcs(template.FuncMap{
```
Comment (alyssacgoins) [inline controllers/config/templating.go]: Why is this change necessary?

File: controllers/config/templating.go [hunk 9]
```diff
@@ -69,3 +73,17 @@ func templateSource(r io.Reader, context interface{}) (mf.Source, error) {
 	}
 	return mf.Reader(&b2), nil
 }
+
+func yamlString(value interface{}) (string, error) {
```
Comment (alyssacgoins) [inline controllers/config/templating.go]: Can you give this function a more descriptive name and add an explanatory comment as well?

File: controllers/resource_claim_validation_test.go [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+	dspav1 "github.com/opendatahub-io/data-science-pipelines-operator/api/v1"
+	"github.com/opendatahub-io/data-science-pipelines-operator/controllers/testutil"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (s *ControllerSuite) TestResourceClaimCRDValidation() {
```
Comment (alyssacgoins) [inline controllers/resource_claim_validation_test.go:33]: Can you add a test case for a nil resource claim? Or is this not a valid edge case?

**Review by alyssacgoins — CHANGES_REQUESTED** [review]
Backwards compatibility/regression testing in DSPO tends to be implicit - if current testing succeeds when a new API field is added, we assume validation. From your perspective here, are there any regression edge cases missing here? Is there any way this change can break exisiting functionality?

File: api/v1/zz_generated.deepcopy.go (hunk 7 above)
Comment (hbelmiro) [reply api/v1/zz_generated.deepcopy.go:41]: This is a generated file.

File: controllers/resource_claim_validation_test.go (hunk 10 above)
Comment (hbelmiro) [reply controllers/resource_claim_validation_test.go:33]: A nil resource claim is not a valid edge case here: `ResourceClaims` is a slice of `PodResourceClaim` values, so individual entries cannot be nil. A nil or empty slice is valid and omitted from rendered manifests. Nil claim-source pointers are covered by the “missing claim source” case and rejected by the CRD validation rule.

File: controllers/config/templating.go (hunk 8 above)
Comment (hbelmiro) [reply controllers/config/templating.go]: No longer needed. Removed.

File: controllers/config/templating.go (hunk 9 above)
Comment (hbelmiro) [reply controllers/config/templating.go]: No longer needed. Removed.

**Review by alyssacgoins — APPROVED** [review]
/lgtm
